tests/core/pthread/sync_to_async.cpp

```diff
@@ -0,0 +1,151 @@
+#include <assert.h>
+#include <emscripten.h>
+#include <pthread.h>
+
+#include <thread>
+#include <functional>
+#include <iostream>
+
+// Helper class for generic sync-to-async conversion. doWork() lets you provide
+// some code to be run, which is done on a helper thread. The thread can do
+// async operations, and it has a callback to indicate when it finishes working.
+// The calling thread mainwhile blocks as it waits.
+//
+// This can be useful if you are a location where blocking is possible (like a
+// thread, or when using PROXY_TO_PTHREAD), but you have code that is hard to
+// refactor to be async, but that requires some async operation (like waiting
+// for a JS event).
+class SyncToAsync {
+public:
+  using Callback = std::function<void()>;
+
+  SyncToAsync() : thread(threadMain, this), childLock(mutex) {
+    // The child lock is associated with the mutex, which takes the lock, and
+    // we free it here. Only the child will lock/unlock it from now on.
+    childLock.unlock();
+  }
+
+  ~SyncToAsync() {
+    quit = true;
+
+    // Wake up the child with an empty task.
+    doWork([](Callback func){
+      func();
+    });
+
+    thread.join();
+  }
+
+  // Run some work on thread. This is a synchronous call, but the thread can do
+  // async work for us. To allow us to know when the async work finishes, the
+  // worker is given a function to call at that time.
+  void doWork(std::function<void(Callback)> newWork) {
+    // Send the work over.
+    {
+      std::lock_guard<std::mutex> lock(mutex);
+      work = newWork;
+      finishedWork = false;
+      readyToWork = true;
+    }
+    condition.notify_one();
+
+    // Wait for it to be complete.
+    std::unique_lock<std::mutex> lock(mutex);
+    condition.wait(lock, [&]() {
+      return finishedWork;
+    });
+  }
+
+private:
+  std::thread thread;
+  std::mutex mutex;
+  std::condition_variable condition;
+  std::function<void(Callback)> work;
+  bool readyToWork = false;
+  bool finishedWork;
+  bool quit = false;
+
+  // The child will be asynchronous, and therefore we cannot rely on RAII to
+  // unlock for us, we must do it manually.
+  std::unique_lock<std::mutex> childLock;
+
+  static void* threadMain(void* arg) {
+    auto* parent = (SyncToAsync*)arg;
+    emscripten_async_call(threadIter, arg, 0);
+    return 0;
+  }
+
+  static void threadIter(void* arg) {
+    auto* parent = (SyncToAsync*)arg;
+    // Wait until we get something to do.
+    parent->childLock.lock();
+    parent->condition.wait(parent->childLock, [&]() {
+      return parent->readyToWork;
+    });
+    auto work = parent->work;
+    parent->readyToWork = false;
+    // Do the work.
+    work([parent, arg]() {
```

Comment:
I'm a little bit unclear why the user-provided function needs to call this lambda to resume, rather than just returning. Like, why wouldn't this work
```
work(); // call user code
parent->finishedWork = true;
....
```

Reply:
Imagine that `work()` queues an async event for 10 seconds in the future. In your suggestion we would be marking `finishedWork` immediately, before those 10 seconds pass. So we wouldn't be waiting synchronously for the thread to finish.

Reply:
On, in other words, the client would stash the resume callback somewhere, and invoke it when it gets its own async callback.

Reply:
Yes, exactly.